docs/yayfzf.1
.\" Manpage for yayfzf
.TH yayfzf 1 "2026-01-06" "2.0.13" "yayfzf manual"

.SH NAME
yayfzf \- interactively search, install, and remove AUR packages using fzf


.SH SYNOPSIS
.B yayfzf [\fIOPTIONS\fR] [\fISEARCH QUERY\fR]


.SH DESCRIPTION
Search, install, and remove packages from the AUR interactively with fzf.


.SH OPTIONS
.TP
.B \-k, \-\-keybindings
Show the list of available keybindings and exit.

.TP
.B \-i, \-\-init-config
Install an example configuration file to the XDG config directory and exit.

.TP
.B \-s, \-\-show-config
Print the example configuration file to stdout and exit.

.TP
.B \-v, \-\-verbose
Show verbose output

.TP
.B \-h
Show the short help text and exit.

.TP
.B \-\-help
Show full help, including keybindings and config file locations, and exit.

.TP
.B \-\-help\-man
Show manpage help and exit.

.TP
.B \-V, \-\-version
Show the yayfzf version and exit.

.SH EXAMPLES
.TP
.B $ yayfzf
Run yayfzf with no search query

.TP
.B $ yayfzf fzf
Search for 'fzf' with yayfzf

.TP
.B $ yayfzf \-h
Show the help menu


.SH KEYBINDINGS

.TP
.B ↑
Move up

.TP
.B ↓
Move down

.TP
.B PGUP
Move up one page

.TP
.B PGDN
Move down one page

.TP
.B HOME
Move to first

.TP
.B END
Move to last

.TP
.B TAB
Select item.

.TP
.B Shift\-TAB
Unselect item.

.TP
.B Ctrl\-a
Select all items.

.TP
.B Ctrl\-d
Deselect all selected items.

.TP
.B ENTER
Perform action on the current selection.

.TP
.B Ctrl\-i
Install selected packages

.TP
.B Ctrl\-r
Uninstall package

.TP
.B Ctrl\-u
Update all packages

.TP
.B Ctrl\-l
Clear the current search query and screen.

.TP
.B Ctrl\-Backspace
Clear the current search query.

.TP
.B Ctrl\-/
Cycle fzf layout and preview window position.

.TP
.B Ctrl\-v
Toggle the preview window.

.TP
.B ?
Show keybindings.

.TP
.B Ctrl\-c
Exit yayfzf.

.TP
.B Ctrl\-w
Exit yayfzf.

.TP
.B ESC
Exit yayfzf.


.SH ENVIRONMENT
.TP
.B YAYFZF_PREVIEW_WINDOW
Controls the layout and size of the fzf preview window.

The value is passed directly to fzf’s
.BR --preview-window
option.

Default:
.RS
.B bottom,43%,border-double
.RE

.TP
.B NO_COLOR
Disable all ANSI color output.

When set, yayfzf disables:
.RS
.IP \(bu 2
fzf color themes
.IP \(bu 2
ANSI color output in messages

.PP
This follows the
.B NO_COLOR
convention used by many CLI tools.
.RE


.SH SEE ALSO
.BR fzf (1)
.BR yay (1)
.BR rclonefzf (1)


.SH AUTHOR
Conner Will <conner.will@connerwill.com>


yayfzf.sh
#!/usr/bin/env bash
#yayfzf2.sh
########################################
#[============== YAYFZF ==============]#
########################################

##-------- OPTIONS --------##
packages_sort_by="votes"
use_color="always"
REQUESTSPLITN=100

#searchFunction="yay --sync --search --quiet --sortby $packages_sort_by $SearchInput" #(not currently used)
detailedPreview="yay --sync --info --info --verbose --color $use_color "
installPackage="yay --sync --verbose --color $use_color --nobatchinstall --confirm"
##-------------------------##

PRODUCTNAME="yayfzf"
VERSION="0.1.2"
LICENSE="GNU General Public License, version 3"
WEBSITE="https://github.com/ConnerWill/yayfzf"
AUTHOR="ConnerWill"
DATE="2022-05-05"



_keybinding_file="lib/KEYBINDINGS.conf"
[[ -e "$_keybinding_file" ]] \
	&& source "$_keybinding_file" \
	|| KEYBINDINGS=""


function _version() {
	cat <<EOS
========================================================
    NAME    :    ${PRODUCTNAME}
    VERSION :    v${VERSION}
    DATE    :    ${DATE}
    LICENCE :    ${LICENSE}
    AUTHOR  :    ${AUTHOR}
    WEBSITE :    ${WEBSITE}/${AUTHOR}
========================================================
EOS
}

function _help() {
	cat <<EOS

    NAME
        yayfzf

    DESCRIPTION
        search and install packages from the AUR with fzf

    USAGE
        yayfzf    KEYWORDS...
        yayfzf    -h|--help
        yayfzf    -V|--version

    KEYBINDINGS
$KEYBINDINGS

========================================================
EOS
}

SHORTKEYBINDINGS="\
Alt+d           Install Package(s)
[Tab/Ctl+Tab]   [Select/Deselect]
Page[Dn/Up]     Page [Down/up]
[Home/End]      [Top/Bottom]
Ctrl+/          Change Layout
?|Ctl+h         Help
Esc|Ctl+q       Exit"

function _short_help() {
	cat <<EOS
$SHORTKEYBINDINGS
EOS
}

##---------------##
### Main Script ###
##---------------##

## Error messege
function _err() {
	printf "\n[\e[31m WARNING \e[m] %s\n" "${*}"
	echo -e -n "\nSearch yay\n\nInput Required!\nPlease Enter Initial Search Query: "
	read -r SearchInput
	[[ -z "$SearchInput" ]] && return 1
	_main "${SearchInput}"
}

## Check if user sent a search query when running.
## If no input received, show warning and the help menu,
## then prompt for input. If input is still empty, Exit.
function _main() {
	if [[ "${#}" -le 0 ]]; then
		_version
		_help
		_err "No Input Received ..."
		return 1
		_main
	fi
	for arg in "${@}"; do
		case "${arg}" in
		-h | --help)
			_help
			return 0
			;;
		-V | --version)
			_version
			return 0
			;;
		esac
	done
	clear
	_search_packagemgr "${*}" | _fzf
}

## Search package manager
function _search_packagemgr() {
	local searchQuery
	searchQuery="${*}"
	yay \
		--sync \
		--search \
		--requestsplitn "$REQUESTSPLITN" \
		--quiet \
		--sortby "$packages_sort_by" \
		--searchby name \
		"$SearchInput"
}

## Main script. Run fzf with package manager search results.
function _fzf() {
	fzf </dev/stdin \
		--query "$SearchInput" \
		--print-query \
		--select-1 \
		--exit-0 \
		--no-clear \
		--extended \
		--tac \
		--no-sort \
		--keep-right \
		--header "${KEYBINDINGS}" \
		--header-first \
		--header-lines=0 \
		--multi \
		--marker="✔ " \
		--prompt="> " \
		--pointer="[>" \
		--info=inline \
		--height=100% \
		--ansi \
		--color='16,bg:#1A1A1A,bg+:#333333,info:#999999,border:#999999,spinner:#F5F5F5' \
		--color='hl+:#BD2C00,hl:#4078C0,fg:#C1C1C1,fg+:#FFFFFF,header:#666666,preview-bg:#333333' \
		--color='header:#666666,query:#4183c4,pointer:#BD2C00,marker:#6CC644,prompt:#F5F5F5' \
		--color='gutter:#141414' \
		--layout 'default' \
		--preview "$detailedPreview {}" \
		--preview-window "right:65%" \
		--margin 0%,0% \
		--padding 3%,2% \
		--border "rounded" \
		--delimiter=" " \
		--tabstop=4 \
		--filepath-word \
		--scroll-off=0 \
		--hscroll-off=100 \
		--expect=ctrl-c,esc \
		--bind "alt-d:execute: echo 'Installing: {+}' && sleep 1 && $installPackage {+} || echo '\nERROR Installing: {+}\nReturn to yayfzf? (Y/n' ; read -s -q || exit ||  echo  'Return to yayfzf? (Y/n)' ; read -s -q || exit" \
		--bind 'pgdn:page-down' \
		--bind 'pgup:page-up' \
		--bind 'home:last' \
		--bind 'end:first' \
		--bind 'ctrl-/:change-preview-window(up,border-rounded|up,40%,border-rounded|left,border-rounded|left,border-rounded,40%|down,border-rounded|down,40%,border-rounded|down,10%,border-rounded|hidden|right,40%,border-rounded|right,70%,border-rounded|right,90%,border-rounded)' \
		--bind 'ctrl-v:toggle-preview' \
		--bind "?:preview: less $SHORTKEYBINDINGS" \
		--bind "f1:preview: echo $SHORTKEYBINDINGS" \
		--bind "ctrl-h:preview: echo $SHORTKEYBINDINGS" \
		--bind 'tab:select' \
		--bind 'shift-tab:deselect' \
		--bind 'ctrl-d:deselect-all' \
		--bind 'alt-left:kill-word' \
		--bind 'alt-bspace:clear-query' \
		--bind 'ctrl-l:clear-query' \
		--bind 'ctrl-q:print-query' \
		--bind "esc:accept-non-empty"
}

function yayfzf() {
	_main "${@}"
}

##---------------##
###  END MAIN   ###
##---------------##


### TODO:
##
##      Add Next/Previous page function
###
#        --bind "ctrl-n:reload(_next_page)"  \
#        --bind "ctrl-p:reload(_previous_page)
#
#   gh_page_number_max=3
#
#    function _next_page(){
#        [[ "$gh_page_number" -lt "$gh_page_number_max" ]] && gh_page_number=$((gh_page_number-1))
#        _main "$query"
#        return
#    }
#
#    function _previous_page(){
#        [[ "$gh_page_number" -gt 1 ]] && gh_page_number=$((gh_page_number-1))
#        _main "$query"
#        return
#    }
###

### TODO
#        --bind "ctrl-i:header ${SHORTKEYBINDINGS}" \
